sbi/utils/posterior_ensemble.py

```diff
@@ -0,0 +1,446 @@
+                `num_to_optimize` with highest log-probability as the initial points
+            learning_rate: Learning rate of the optimizer.
+            init_method: How to select the starting parameters for the optimization. If
+                it is a string, it can be either [`posterior`, `prior`], which samples
+                the respective distribution `num_init_samples` times. If it is a
+                tensor, the tensor will be used as init locations.
+            num_init_samples: Draw this number of samples from the posterior and
+                evaluate the log-probability of all of them.
+                for the optimization.
+            save_best_every: The best log-probability is computed, saved in the
+                `map`-attribute, and printed every `save_best_every`-th iteration.
+                Computing the best log-probability creates a significant overhead
+                (thus, the default is `10`.)
+            show_progress_bars: Whether or not to show a progressbar for sampling from
+                the posterior.
+            individually: If true, returns log weights and MAPs individually.
+
+        Returns:
+            The ensemble MAP estimate or individual log_weigths and component MAP
+            estimate if individually == True.
+        """
+
+        if individually:
+            maps = []
+            log_weights = torch.log(self._weights).reshape(-1, 1)
+
+            for posterior in self.posteriors:
+                maps.append(
+                    posterior.map(
+                        x=x,
+                        num_iter=num_iter,
+                        num_to_optimize=num_to_optimize,
+                        learning_rate=learning_rate,
+                        init_method=init_method,
+                        num_init_samples=num_init_samples,
+                        save_best_every=save_best_every,
+                        show_progress_bars=show_progress_bars,
+                    )
+                )
+            maps = torch.stack(maps)
+            return log_weights, maps
+
+        else:
+            if init_method == "posterior":
+                inits = self.sample((num_init_samples,))
+            elif init_method == "proposal":
+                inits = self.proposal.sample((num_init_samples,))
+            elif isinstance(init_method, Tensor):
+                inits = init_method
+            else:
+                raise ValueError
+
+            return gradient_ascent(
+                potential_fn=self.potential_fn,
+                inits=inits,
+                theta_transform=self.theta_transform,
+                num_iter=num_iter,
+                num_to_optimize=num_to_optimize,
+                learning_rate=learning_rate,
+                save_best_every=save_best_every,
+                show_progress_bars=show_progress_bars,
+            )[0]
+
+
+class EnsemblePotentialProvider(BasePotential):
```

Comment:
As of #573, we are no longer using the term `provider`. I suggest calling this class `EnsemblePotential`